platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceRegistration.java

```diff
@@ -34,6 +34,14 @@ public interface ServiceRegistration {
      */
     void add(Class<?> serviceType);
 
+    /**
+     * Adds a service to this registry. The implementation class should have a single public constructor, and this constructor can take services to be injected as parameters.
+     *
+     * @param serviceType The service to make visible.
+     * @param implementationType The implementation type of the service.
+     */
+    void add(Class<?> serviceType, Class<?> implementationType);
```

Comment:
💭  A new internal API to register a lazy service implementation for a given service type.